### ISSUE TITLE
On device ad blocking does not work under DNSCrypt when app auto starts after reboot

### ISSUE THREAD

**Issue by modestderp** [celzero/rethink-app#326]
I noticed under the following setup, while RethinkDNS auto starts successfully after device reboot, it fails to block target traffic.

Device: Samsung S10
OS: Android 10
DNS: DNSCrypt, using DNSCrypt servers with DNSSEC enabled
Additional info: Using on-device blocklists

Repro steps:
1. Download blocklists to device and enable target blocklists.
2. Enable DNSCrypt instead of DoH.
3. Select DNSCrypt server.
4. Enable "auto-start on power-up (experimental)" option.
5. Restart device.
6. Observe that RethinkDNS auto starts itself after reboot on top bar.
7. Ensure RethinkDNS is running in app homescreen. (i.e. "PROTECTED")
8. Launch any app to trigger blocklists (e.g. Facebook).
9. Observe logs. The requests to target requests are NOT blocked.

So far the only way to solve this is to change DNSCrypt back to DoH (e.g. RethinkDNS Plus), and then back to DNSCrypt as a sort of "reset". This usually will re-enforce the local blocklist on the traffic once again.

Would appreciate if this issue could be investigated. Thanks.

**Comment by ignoramous** [comment]
Thanks for the detailed bug report. We likely have enough information to debug this. Just a couple more questions:

Is it 100% reproducible?

Are you using a custom DNSCrypt resolver? If so, is it public? If yes, can you please share its dnsstamp?

**Comment by modestderp** [comment]
Yes, it is 100% reproducible on my side. Funny enough, I encounter a similar issue when I manually stop the app under DNSCrypt and then manually start again. I need to switch to DoH and then back to DNSCrypt to "reset".

Mainly I use the following public DNSCrypt servers:
**jp.tiar.app**
sdns://AQcAAAAAAAAAEjE3Mi4xMDQuOTMuODA6MTQ0MyAyuHY-8b9lNqHeahPAzW9IoXnjiLaZpTeNbVs8TN9UUxsyLmRuc2NyeXB0LWNlcnQuanAudGlhci5hcHA

**saldns01-conoha-ipv4**
sdns://AQcAAAAAAAAAFDExOC4yNy4xMDguMTQwOjUwNDQzIJ6mPj1aeYrSINYeUzVU2n9JMTUgalCmDR6OhSlzrFtfIjIuZG5zY3J5cHQtY2VydC5zYWxkbnMwMS50eXBlcS5vcmc

Basically any non-logging, non-censoring DNSSEC public DNS resolver in this list.
[https://dnscrypt.info/public-servers/](https://dnscrypt.info/public-servers/)

**Comment by ignoramous** [comment]
Okay, I can reproduce this 100% of the time too. Marking this issue highest-priority. Taking a look.

**Comment by ignoramous** [comment]
Local blocklists are set only when DNS mode is either [`None`](https://github.com/celzero/firestack/blob/e3ae8df8cfe62d51fe77710aa1641a736b4154dd/intra/settings/config.go#L16), `IP`, `Port` (that is, when it's a DoH endpoint).

https://github.com/celzero/rethink-app/blob/987570b44fc6935db93ef9f8096b4daf3a5435d7/app/src/main/java/com/celzero/bravedns/net/go/GoVpnAdapter.java#L187

Seems like a miss.

**Comment by ignoramous** [comment]
Implemented, yet-to-release it.
